llama_stack/providers/inline/vector_io/sqlite_vec/sqlite_vec.py

```diff
@@ -115,35 +130,69 @@ async def add_chunks(self, chunks: List[Chunk], embeddings: NDArray, batch_size:
         finally:
             cur.close()  # Ensure cursor is closed
 
-    async def query(self, embedding: NDArray, k: int, score_threshold: float) -> QueryChunksResponse:
+    async def query(
+        self, embedding: Optional[NDArray], query_str: Optional[str], k: int, score_threshold: float, search_mode: str
+    ) -> QueryChunksResponse:
         """
+        Supports both vector-based and keyword-based searches.
+
+        1. Vector Search (`search_mode`="vector")
         Query for the k most similar chunks. We convert the query embedding to a blob and run a SQL query
         against the virtual table. The SQL joins the metadata table to recover the chunk JSON.
+
+        2. Keyword search (`search_mode`="keyword")
+        Uses SQLite's `FTS5` module to find matches for `query_str` and ranks results using BM25 relevance scoring.
         """
-        emb_list = embedding.tolist() if isinstance(embedding, np.ndarray) else list(embedding)
-        emb_blob = serialize_vector(emb_list)
         cur = self.connection.cursor()
-        query_sql = f"""
-            SELECT m.id, m.chunk, v.distance
-            FROM {self.vector_table} AS v
-            JOIN {self.metadata_table} AS m ON m.id = v.id
-            WHERE v.embedding MATCH ? AND k = ?
-            ORDER BY v.distance;
-        """
-        cur.execute(query_sql, (emb_blob, k))
+        if search_mode == "vector":
+            emb_list = embedding.tolist() if isinstance(embedding, np.ndarray) else list(embedding)
+            emb_blob = serialize_vector(emb_list)
+
+            query_sql = f"""
+                SELECT m.id, m.chunk, v.distance
+                FROM {self.vector_table} AS v
+                JOIN {self.metadata_table} AS m ON m.id = v.id
+                WHERE v.embedding MATCH ? AND k = ?
+                ORDER BY v.distance;
+            """
+            cur.execute(query_sql, (emb_blob, k))
+        elif search_mode == "keyword":
+            if query_str is None:
+                raise ValueError("query_str is required for keyword search.")
+            # Full-Text Search SQL
+            query_sql = f"""
+                SELECT DISTINCT m.id, m.chunk, bm25({self.fts_table}) AS score
```

Comment:
I think `DISTINCT` here makes sense, since we would not want to repeat the output with the same chunk ids. 
Thoughts: @franciscojavierarceo?

Reply:
Yeah correct.